src/main/kotlin/org/jitsi/nlj/util/ArrayCache.kt

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright @ 2019 - present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.nlj.util
+
+import org.jitsi.utils.TimeProvider
+import java.lang.Integer.max
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * Implements a fixed-sized cache based on a pre-filled array. The main use-case is the outgoing RTP packet cache.
+ *
+ * @author Boris Grozev
+ */
+open class ArrayCache<T>(
+    val size: Int,
+    val synchronize: Boolean = false,
+    /**
+     * The function to use to clone items. The cache always saves copies of the items that are inserted.
+     */
+    private val cloneItem: (T) -> T,
+    private val timeProvider: TimeProvider = TimeProvider()
+) {
+    private val cache: Array<Container> = Array(size) { Container() }
+    protected val syncRoot = Any()
+    /**
+     * The index in [cache] where the item with the highest index is stored.
+     */
+    private var head = -1
+
+    var numInserts = 0
+    var numOldInserts = 0
+    private val _numHits = AtomicInteger()
+    private var _numMisses = AtomicInteger()
+    val numHits
+        get() = _numHits.get()
+    val numMisses
+        get() = _numMisses.get()
+    val hitRate = _numHits.get() * 1.0 / max(1, _numHits.get() + _numMisses.get())
+
+    protected val lastIndex: Int
+        get() = if (head == -1) -1 else cache[head].index
+
+    /**
+     * Inserts an item with a specific index in the cache. Stores a copy.
+     */
+    fun insertItem(item: T, index: Int): Boolean =
+        if (synchronize) {
```

Comment:
was this for a testing comparison? do still want to keep both ways?

Reply:
I was hoping to get away without synchronization, but I don't think it's gonna work out (we have one thread inserting, another one reading, and a third one doing getMany). Still, if we end up using it somewhere else (e.g. the frame projection stuff) we may be able to not synchronize

Reply:
gotcha.  maybe we should default `synchronize` to true and put it after the `cloneItem` arg? that way we can use the default without having to use named args